core/shared/src/main/scala/spire/math/poly/Monomial.scala

```diff
@@ -0,0 +1,375 @@
+
+  def x[@spec(Double) C: ClassTag: Order](implicit r: Ring[C]): Monomial[C] =
+    new Monomial[C](r.one, SortedMap('x' -> 1))
+
+  object Re {
+    val variable = "[\u03B1-\u03C9a-zA-Z]".r
+    val numFractional = "[0-9]+(?:[/.][0-9]+)?".r
+    val numIntegral = "[0-9]+".r
+    val varExp = s"($variable)(?:\\^(-?[0-9]+))?".r
+
+    def term(number: Regex): Regex = s"(- *)?($number)?($varExp)*".r
+    val termFractional = term(numFractional)
+    val termIntegral = term(numIntegral)
+  }
+
+
+    // val terms = s"($negative)? *($number)?($varchar(\\^(-)?$number)?)*".r
+
+  final def parseFractional[@spec(Double) C: ClassTag: Fractional](str: String): Monomial[C] = {
+
+    val f = implicitly[Field[C]]
+    val c = implicitly[ConvertableTo[C]]
+
+    val stringToC: String => C = (str: String) => str.split('/') match {
+      case (Array(numerator, denominator)) =>
+        c.fromDouble(numerator.toDouble) / c.fromDouble(denominator.toDouble)
+      case (Array(number)) => c.fromDouble(number.toDouble)
+      case _ => throw new Exception("Could not parse number " + str)
+    }
+
+    parseGeneric(str, Re.termFractional, stringToC, f.one)
+  }
+
+  final def parseIntegral[@spec(Int, Long) C: Integral: ClassTag](str: String): Monomial[C] = {
+
+    val r = implicitly[Integral[C]]
+    val c = implicitly[ConvertableTo[C]]
+
+    val stringToC: String => C = (str: String) => c.fromBigInt(BigInt(str))
+
+    parseGeneric(str, Re.termIntegral, stringToC, r.one)
+  }
+
+  private [poly] final def parseGeneric[@spec(Double, Int, Long) C: ClassTag: Order: ConvertableTo: Ring](str: String, terms: Regex, stringToC: String => C, one: C): Monomial[C] = {
+
+    val m: Option[Monomial[C]] = terms.findFirstMatchIn(str).map { term =>
+      val coeff: C = term.subgroups.map(Option(_)) match {
+        case (List(Some(neg), Some(coeff), _, _, _)) => -stringToC(coeff)
+        case (List(None, Some(coeff), _, _, _)) => stringToC(coeff)
+        case (List(Some(neg), None, _, _, _)) => -one
+        case (List(None, None, _, _, _)) => one
+        case _ => one
+      }
+      val exponents = Re.varExp.findAllMatchIn(term.matched).toList.map {
+        _.subgroups match {
+          case (List(variable, null)) => variable(0) -> 1
+          case (List(variable, exp)) => variable(0) -> exp.toInt
+          case _ => throw new Exception("Couldn't parse term " + term)
+        }
+      }
+      Monomial[C](coeff, exponents)
+    }
+    m getOrElse(throw new Exception("Could not parse term "+str))
+  }
+
+  private val IsZero = "0".r
+  private val IsNegative = "-(.*)".r
+
+  implicit def monomialEq[@spec(Double) C: ClassTag: Order: Semiring] = new MonomialEq[C] {
+    val scalar = Semiring[C]
+    val ct = implicitly[ClassTag[C]]
+  }
+
+}
+
+// An equivalent monomial has the same variables (that's all!)
+// not checking that the variable exponents are equal using this instance
+trait MonomialEq[@spec(Double) C] extends Eq[Monomial[C]] {
```

Comment:
What's the motivation of an `Eq` instance that does not obey the `Eq` laws ? If there is no other solution, we have to hide this and make it private[something].